File: ReadingTracker/app/src/main/java/com/codelab/readingtracker/reading/SharedPreferencesEntryProvider.java
package com.codelab.readingtracker.reading;

import android.content.Context;
import android.content.SharedPreferences;

import androidx.preference.PreferenceManager;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Objects;
import java.util.Set;

public class SharedPreferencesEntryProvider implements EntryProvider {

    private SharedPreferences sharedPreferences;

    public SharedPreferencesEntryProvider(Context context) {
        sharedPreferences = PreferenceManager.getDefaultSharedPreferences(Objects.requireNonNull(context));

    }

    @Override
    public void save(List<Entry> entries) {
        List<String> strings = new ArrayList<>();
        entries.forEach((entry -> strings.add(entry.toString())));
        Set<String> set = new HashSet<>(strings);

        SharedPreferences.Editor editor = sharedPreferences.edit();
        editor.remove("Entries");
        editor.putStringSet("Entries", set).apply();
    }

    @Override
    public List<Entry> load() {
        List<Entry> entries = new ArrayList<>();

        new ArrayList<>(sharedPreferences.getStringSet("Entries", new HashSet<>())).forEach((s) -> entries.add(Entry.fromString(s)));

        return entries;
    }
}

File: ReadingTracker/app/src/main/java/com/codelab/readingtracker/reading/SharedPreferencesEntryProvider.kt
package com.codelab.readingtracker.reading

import android.content.Context
import android.content.SharedPreferences

import androidx.preference.PreferenceManager

import java.util.ArrayList
import java.util.HashSet
import java.util.Objects

class SharedPreferencesEntryProvider(context: Context) : EntryProvider {

    private val sharedPreferences: SharedPreferences = PreferenceManager.getDefaultSharedPreferences(Objects.requireNonNull(context))

    override fun save(entries: List<Entry>) {
        val strings = ArrayList<String>()
        entries.forEach { entry -> strings.add(entry.toString()) }
        val set = HashSet(strings)

        val editor = sharedPreferences.edit()
        editor.remove("Entries")
        editor.putStringSet("Entries", set).apply()
    }

    override fun load(): List<Entry> {
        val entries = ArrayList<Entry>()

        ArrayList(sharedPreferences.getStringSet("Entries", HashSet())!!).forEach { s -> entries.add(Entry.fromString(s)) }

        return entries
    }
}
